proposals/NNNN-swiftpm-additional-plugin-apis.md

```diff
@@ -0,0 +1,450 @@
+
+This is the same structure that SwiftPM’s built-in subsystems currently use, and the intent is that, over time, at least some of those subsystems can be reimplemented as plugins. This information is also expected to be useful to all kinds of plugins.
+
+In addition to the new information, this proposal adds API for traversing the package graph, such as being able to access topologically sorted lists of target dependencies. This will make it more convenient for build tool plugins that need to, for example, generate command line arguments that include a search paths for each dependency target.
+
+## Detailed Design
+
+This proposal defines a new `PluginContext` structure that contains:
+
+* a reference to the `Package` object at the root of the subgraph to which the plugin is being applied
+* the contextual information that was previously part of `TargetBuildContext`
+
+This structure factors out all the information related to the package graph, such as the package and target names and directories, leaving the context with just the top-level contextual information.
+
+The `BuildToolPlugin` protocol entry point defined by SE-0303 is superseded with a new entry point that takes the new `PluginContext` type and a reference to the `Target` for which build commands should be generated, but the previous API remains so that existing plugins continue to function.
+
+### Plugin API
+
+The new `PluginContext` structure in the `PackagePlugin` API is defined as:
+
+```swift
+/// Provides information about the package for which the plugin is invoked,
+/// as well as contextual information based on the plugin's stated intent
+/// and requirements.
+public struct PluginContext {
+    /// Information about the package to which the plugin is being applied.
+    public let package: Package
+
+    /// The path of a writable directory into which the plugin or the build
+    /// commands it constructs can write anything it wants. This could include
+    /// any generated source files that should be processed further, and it
+    /// could include any caches used by the build tool or the plugin itself.
+    /// The plugin is in complete control of what is written under this di-
+    /// rectory, and the contents are preserved between builds.
+    ///
+    /// A plugin would usually create a separate subdirectory of this directory
+    /// for each command it creates, and the command would be configured to
+    /// write its outputs to that directory. The plugin may also create other
+    /// directories for cache files and other file system content that either
+    /// it or the command will need.
+    public let pluginWorkDirectory: Path
+
+    /// The path of the directory into which built products associated with
+    /// the target are written.
```

Comment:
which target?

Reply:
Good point — this was a leftover from the `TargetBuildContext`, and this actually points out that this will not scale well for other kinds of plugins.  I’m going to have to take a look at what should be done about this one — the build tool plugins used this for forming some paths but perhaps there should instead be an API to look this up per-target.  I’ll have to think about this more and revise the proposed API here.